### ISSUE TITLE
Restarting Jenkins with selenium nodes on Windows slaves causes duplicate nodes to be started on Windows slaves

### ISSUE THREAD

**Issue by zdmeyers** [jenkinsci/selenium-plugin#68]
If I restart my entire jenkins server (like, after updating a plugin), the Windows slaves I have in my Grid setup will show up as having 2 Selenium nodes running on them (one on port 4445 and one on port 5555). I have to manually go onto these slaves, kill the jenkins slave process, then restart it by hand from the slaves in order for these duplicates to go away. I am running the Windows slaves from a .bat file (started by Task Scheduler) that connects over JNLP. One of the Windows slaves is Win7, the other is Win8.1

**Comment by Mobrockers** [comment]
What plugin version are you on? If you updated to version 2.53.0, did this issue happen on 2.4.1?
Could you also take some screenshots to explain exactly what is happening on the slaves, and attach the node configurations?

It kinda sounds like you might have two node configurations that match on both windows slaves, thus creating two selenium nodes on both slaves.

**Comment by zdmeyers** [comment]
![image](https://cloud.githubusercontent.com/assets/8092276/14753169/221dc792-0899-11e6-905e-cb708c3b14f5.png)

![image](https://cloud.githubusercontent.com/assets/8092276/14753237/94132914-0899-11e6-9769-627c9cbc3dbf.png)

![image](https://cloud.githubusercontent.com/assets/8092276/14753276/de7f0fd6-0899-11e6-84a3-cecfd1c36d39.png)

![image](https://cloud.githubusercontent.com/assets/8092276/14753286/f1380b1e-0899-11e6-80c5-b84131b2b095.png)

Each node has a label attached to it, even the Linux and Solaris ones. Don't mind the redactions, these are production environments from my company and I cannot give out IPs and the like. I have left the relevant information visible. The second image, the two sets of arrows show the duplicates. One is running on port 4445 and the other on 5555 on two different machines (first set has ie10 as the browser and second set has ie11).

The last two images show the configurations as they appear on the UI.

**Comment by zdmeyers** [comment]
And yes, it has happened ever since I installed the plugin (i believe the first version I used was 2.4.1)

**Comment by Mobrockers** [comment]
Right I get it. Only master is restarted so when master restarts, the plugin automatically starts a selenium node for you on every slave, but since your slaves were not restarted the existing selenium nodes are still there and reconnect to the selenium hub automatically (see https://github.com/seleniumhq/selenium/issues/1573).

So I should find out whether the plugin currently catches a jenkins shutdown and what it does in this case, and I should either fix or extend this so that nodes on slaves are closed while jenkins is preparing for shutdown. Agreed?

**Comment by Mobrockers** [comment]
@zdmeyers Would you mind testing a change I've made that I think should fix this issue? I don't really have access to the hardware required to test this setup, and setting it up in vm's would if I'm honest take too much of my time right now.

Here's a link to the hpi file: https://dl.dropboxusercontent.com/u/5483751/selenium.hpi
You can install it on your server from the advanced tab on the manage plugins page. If you do not wish to test this for me I'll try to set up my own cluster maybe next weekend if I have time.

**Comment by zdmeyers** [comment]
I will give this version a try when I get into the office tomorrow. Can you tell me what you did to it that you think might fix it?

**Comment by Mobrockers** [comment]
I added a restart listener that stops all nodes on restart (if it works).

**Comment by Mobrockers** [comment]
Currently it listens for shutdown and startup of "computers", but since the slaves never go offline, their computers never show as offline, so selenium grid plugin doesn't stop the nodes on those computers. I'm guessing. I'm brand new to plugin development honestly so still figuring all this stuff out.

**Comment by zdmeyers** [comment]
Im just glad someone picked this plugin up as the original developer seems to have left the project and I never thought I would get this bug looked at

**Comment by Mobrockers** [comment]
As far as I can tell the original developer is kohsuke, the creator of
Jenkins, so I'm not surprised he has more important things to do right now
:p I have heard from darkrift (the other listed maintainer and the one
who's done the most work) that he is not able to do work on this plugin at
his day job so understandably he has not had much time in the past years to
work on it. It is how it goes with such projects I suppose.

**Comment by zdmeyers** [comment]
Your change doesn't seem to have worked. Installed the build you gave me, restarted, then restarted again and still have duplicates.

**Comment by Mobrockers** [comment]
@zdmeyers Mhm. Do you keep getting more and more duplicates each restart? Can you confirm that the nodes are connected in the selenium grid console as well? As in, not through the plugin but rather through selenium grid's own config page. You should be able to get there from the status page in the plugin.

**Comment by zdmeyers** [comment]
![image](https://cloud.githubusercontent.com/assets/8092276/14786817/3acb6b4e-0ac6-11e6-8784-129ae0030b3d.png)
Nope, only 1 duplicate per Windows slave (2 windows slaves = 2 duplicates). And it is only for nodes on Windows slaves. Linux/Solaris are perfectly fine after a restart.

**Comment by Mobrockers** [comment]
Are the configurations exactly the same?

**Comment by zdmeyers** [comment]
![image](https://cloud.githubusercontent.com/assets/8092276/14787047/5d834836-0ac7-11e6-9259-0eb66c56eff7.png)
![image](https://cloud.githubusercontent.com/assets/8092276/14787106/a24fdcae-0ac7-11e6-9f2e-3572efbc0f32.png)

All of the configs are the same except the ports. Each duplicate has the same remoteHost, hubHost, hub, and host. Also all of the other settings are the same.

**Comment by Mobrockers** [comment]
You forgot to black out some info on the other picture so I have removed it for you

**Comment by zdmeyers** [comment]
Oops, thanks

**Comment by remcowesterhoud** [comment]
When you kill the Jenkins slave, do both nodes quit or only one of them?

**Comment by zdmeyers** [comment]
Both nodes quit upon killing the slave process (or rebooting the slave)

**Comment by Mobrockers** [comment]
When the slave is restarted, do any nodes reconnect to the hub? If so on what port do they run?

**Comment by zdmeyers** [comment]
If I manually restart the slave, only 1 node gets started on port 4445. That is how I have been fixing this problem. Log onto slave -> kill slave process -> wait for Grid to recognize that the nodes are gone and remove them -> restart slave by running batch script to connect over JNLP -> Only 1 node gets started

**Comment by Mobrockers** [comment]
I honestly cannot imagine why windows would be a special case so I'll have to try and replicate this behaviour before I can investigate more, sorry.

**Comment by zdmeyers** [comment]
No problem. It is more of an annoyance than anything really program breaking

**Comment by zdmeyers** [comment]
So I have some more information on this. I just installed the new plugin version (2.53.1) and can confirm that it still exists. However, I took a look at one of the Windows slave's log file and saw this:

<===[JENKINS REMOTING CAPACITY]===>Slave.jar version: 2.53.3
This is a Windows slave
Installing chromedriver to C:\Jenkins\tools\chromedriver
Starting Selenium nodes on oaklx8658-vm1
Creating selenium node VM
$ "C:\Program Files\Java\jre1.8.0_73\bin\java" -Dwebdriver.chrome.driver=C:\Jenkins\tools\chromedriver\chromedriver.exe -Dwebdriver.ie.driver=C:\Jenkins\IEDriverServer.exe -cp C:\Users\root\AppData\LocalLow\Sun\Java\Deployment\cache\6.0\4\598455c4-277892d2 hudson.remoting.Launcher -cp C:\Jenkins\selenium-server-standalone-2.53.0.jar;C:\Jenkins\htmlunit-driver-standalone-2.20.jar -connectTo localhost:54799
Starting the selenium node process. Args: [-role, node, -hub, <redacted hub hostname>, -browser, seleniumProtocol=WebDriver,browserName=internet explorer,maxInstances=1,version=10, -browser, seleniumProtocol=WebDriver,browserName=chrome,maxInstances=1, -host, oaklx8658-vm1]
channel started
Slave successfully connected and online
Jun 10, 2016 10:01:55 AM org.openqa.grid.common.RegistrationRequest addCapabilityFromString
INFO: Adding seleniumProtocol=WebDriver,browserName=internet explorer,maxInstances=1,version=10
Jun 10, 2016 10:01:55 AM org.openqa.grid.common.RegistrationRequest addCapabilityFromString
INFO: Adding seleniumProtocol=WebDriver,browserName=chrome,maxInstances=1
10:01:56.468 INFO - Java: Oracle Corporation 25.73-b02
10:01:56.469 INFO - OS: Windows 7 6.1 amd64
10:01:56.474 INFO - v2.53.0, with Core v2.53.0. Built from revision 35ae25b
10:01:56.527 INFO - Driver class not found: com.opera.core.systems.OperaDriver
10:01:56.527 INFO - Driver provider com.opera.core.systems.OperaDriver is not registered
10:01:56.536 INFO - Driver provider org.openqa.selenium.safari.SafariDriver registration is skipped:
registration capabilities Capabilities [{browserName=safari, version=, platform=MAC}] does not match the current platform VISTA
10:01:56.580 INFO - Version Jetty/5.1.x
10:01:56.581 INFO - Started HttpContext[/selenium-server,/selenium-server]
10:01:56.583 INFO - Started org.openqa.jetty.jetty.servlet.ServletHandler@415940e0
10:01:56.583 INFO - Started HttpContext[/wd,/wd]
10:01:56.584 INFO - Started HttpContext[/selenium-server/driver,/selenium-server/driver]
10:01:56.584 INFO - Started HttpContext[/,/]
10:01:56.586 INFO - Started SocketListener on 0.0.0.0:5555
10:01:56.586 INFO - Started org.openqa.jetty.jetty.Server@34983f41
10:01:56.615 INFO - Starting auto registration thread. Will try to register every 5000 ms.
10:01:56.615 INFO - Registering the node to the hub: <redacted hub hostname>
10:01:56.634 INFO - The node is registered to the hub and ready to use

It isn't even recognizing that there is already a node running (in this case, it is running on port 4445) and tries to start a new one on port 5555. The same is true for my other Windows slave. Hopefully this sheds more light on what could potentially be the problem

**Comment by Mobrockers** [comment]
It is not weird that it doesn't recognize another node running because there is no way to actually check that. What is weird is that the node on your windows slave is not stopping when selenium stops, because as far as I know that is what should be happening. The new node is started on port 5555 because when the chosen port (4445) is already taken, the -port option is simply not added to the node launch command.

**Comment by Mobrockers** [comment]
The -port option not being used to launch the node can be seen in this part of your log: "Starting the selenium node process. Args: [-role, node, -hub, (redacted), -browser, seleniumProtocol=WebDriver,browserName=internet explorer,maxInstances=1,version=10, -browser, seleniumProtocol=WebDriver,browserName=chrome,maxInstances=1, -host, (redacted)**(missing -port arg here) **]"

**Comment by zdmeyers** [comment]
Could it be a timing thing?

Jenkins begins to shut down -> plugin stops (but somehow the check to shutdown running nodes isn't run in time/is missed entirely/something) because the Jenkins server has totally gone down?

**Comment by Mobrockers** [comment]
No, jenkins provides a hook on shutdown where shutdown can be interrupted until my hook has completed execution.

**Comment by zdmeyers** [comment]
Hmm. Weird that is isn't reporting that another process is already using port 4445 in the log output. I think if you run selenium standalone from command line and it hits an already used port, it tells you about it.

**Comment by zdmeyers** [comment]
I might not be understanding how those hooks work, but looking at the class you wrote in the above commit, it isn't called anywhere in the code (the RestartListenerImpl class).

**Comment by Mobrockers** [comment]
We're checking if the port is in use before starting the node, without using selenium, so selenium will never throw that exception. I thought there was a log message to indicate this though.

**Comment by Mobrockers** [comment]
It is a hook, extension point, this means jenkins always calls this code itself. This means once you extend it jenkins executes your code when it reaches this part of its own code.

**Comment by zdmeyers** [comment]
That makes sense. But no, I have never seen a log message for port already in use from this plugin.

**Comment by zdmeyers** [comment]
I see how it works now. Excuse my ignorance.

**Comment by Mobrockers** [comment]
That's alright, I don't mind. It should have worked without that hook anyway, because we are already implementating the computerlistener hook, which runs the node shutdown code when the slave goes offline.

**Comment by zdmeyers** [comment]
Quick question, though. Looking at Jenkins implementation of RestartListener, the isReadyToRestart() function JavaDoc says
`    /**
     * Called periodically during the safe restart.
     *
     * @return false to block the restart
     */`

If I do not use the /safeRestart URL (and restarting from upgrade plugin page is done in safe manner or no?) then your override of isReadyToRestart() will never be called. At least, thats what it looks like to me. Should you also override the onrestart() function of the RestartListener class. That one's JavaDoc says: 
`    /**
     * Called immediately before the restart is actually triggered.
     */`

**Comment by zdmeyers** [comment]
I should probably stop talking, lol. Looking further in the class, the isAllReady() function will call isReadyToRestart() for all RestartListener objects returned by the all() function.

**Comment by Mobrockers** [comment]
Mhm, you might have a point there. I'll build you a version with that override soon.

**Comment by zdmeyers** [comment]
I cna submit a PR if you would like?

**Comment by Mobrockers** [comment]
Can't hurt to check though, no harm in talking about it

**Comment by Mobrockers** [comment]
I'm home in a couple minutes and can build you the plugin, it's quicker that way

**Comment by zdmeyers** [comment]
No problem. Thanks for your continued support of this extremely weird (but highly annoying) problem.

**Comment by Mobrockers** [comment]
I will debug this first to see what it actually does, because I might have identified another issue that could cause this. I've checked some other plugin implementations and they do seem to do onRestart, but technically onRestart and isReadyToStart should both work. I'll try onRestart for now.

**Comment by zdmeyers** [comment]
Sounds good. Can you share details about the other issue?

**Comment by Mobrockers** [comment]
The first issue I see is that `new ServiceManagementAction(c).doStop(null);` is called, which should actually throw a nullpointer because `doStop()` calls `getConfiguration(String name)` with null, and `getConfiguration` returns all configurations matching the name. But null is being passed.. So none of this could ever have worked.

**Comment by zdmeyers** [comment]
Good catch. The catch(Throwable e) caluse in the try block will catch the NPE and eat it because it is of type Throwable.

**Comment by Mobrockers** [comment]
ComputerListenerImp.onOffline is also useless because that is called after a slave going offline. I'd buy a beer to person who manages to stop program execution when the slave the program is running on is no longer online.

**Comment by zdmeyers** [comment]
Evil bit-level hacking...

(I will buy you a beer if you recognize the reference without Google :P

**Comment by Mobrockers** [comment]
I could not place the reference without google, no :p 

I did however think of a potential fix. Try this: https://dl.dropboxusercontent.com/u/5483751/selenium.hpi

**Comment by zdmeyers** [comment]
I will install it now and let you know.

**Comment by Mobrockers** [comment]
Now I need to find a way to automatically shut down selenium nodes on jenkins slaves when the jenkins master goes offline (eg kill command or server restart). I don't seem to be able to find anything to do that from the node perspective..

**Comment by zdmeyers** [comment]
Installed and cleaning up duplicates. Once I bring the nodes back online to fix the issue, I will restart jenkins and see if it works.

**Comment by zdmeyers** [comment]
FIXED

**Comment by zdmeyers** [comment]
Whatever it is you did, the duplicates are gone after restarting from the good state

**Comment by Mobrockers** [comment]
```
@Override
    public void onRestart() {

        for (Computer c : PluginImpl.getPlugin().getComputers().keySet()) {
// relevant part
            for (SeleniumGlobalConfiguration conf : PluginImpl.getPlugin().getGlobalConfigurationForComputer(c)) {

                new ServiceManagementAction(c).doStop(conf.getName());
            }
        }
    }
```

**Comment by zdmeyers** [comment]
It was probably both the NPE issue and the overriding of the other function. Although, for safe measure, I would put the same logic in the other function you overrode as well.

**Comment by Mobrockers** [comment]
The isReadyToStart by itself would probably work just the same, except it would hold shutdown until some condition is true, which isn't really necessary for this plugin. It is an abstract method in the base class so it has to be implemented, other than that it's not really needed in this case.

I am going to have to figure out how to deal with unexpected shutdowns, because if your master dies for some reason, and your slaves do not, the same thing issue would occur.

The port is already being logged:

CustomWdConfiguration.java:
```
try {
            opt.addOptionIfSet("-port", c.getChannel().call(new RetrieveAvailablePort(getPort())));
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "an error occured, not adding the port option", e);
        }
```

**Comment by zdmeyers** [comment]
I have no idea how you are going to detect a dead master server from the slave. Only way I can think of is to a listener sitting on the node and pinging the grid hub/master node to see if it is alive. If not, shutdown the selenium node. But I don't think thats possible from a plugin standpoint.

**Comment by Mobrockers** [comment]
The slaves obviously know if they can connect to the master or not, so I would expect to be able to listen for that event, but I can't find anything. I'm going to have to ask on the developer mailing list I guess.

**Comment by zdmeyers** [comment]
If the master is dead, the Selenium plugin is no longer running I assume. So how can you listen for the event from the Slave if your plugin isn't running?

**Comment by Mobrockers** [comment]
I suppose I would want to tell the slave to clean up the processes I created when it loses connection to master.

**Comment by zdmeyers** [comment]
That would be your best bet if it's possible.

**Comment by Mobrockers** [comment]
Anyway, I'll release a new version when I have an answer/fix to unexpected master connection loss. With or without changes to fix that, depending on if it is possible. You can probably keep using the snapshot I sent you without issues and I'll close this issue for now. Please re-open when for some reason the problem comes back.
